scipy/sparse/_sputils.py

```diff
@@ -362,9 +376,13 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
         else:
             raise ValueError('can only specify one unknown dimension')
 
-    if len(new_shape) != 2 and not (allow_1d and len(new_shape) == 1):
+    ndim_new = len(new_shape)
+    if not allow_1d and ndim_new != 2:
         raise ValueError('matrix shape must be two-dimensional')
 
+    if ndim_new >= 3 and allow_1d and not allow_nd:
+        raise ValueError('matrix shape must be one- or two-dimensional')
```

Comment:
While we're at it, can you remove the word `matrix` from the error message (also 2 lines up)? This was pre-existing, but we have arrays now and not only matrices!

Reply:
Changed at https://github.com/scipy/scipy/pull/21197/commits/839d462fb775caac9cffc3a842824a7bdd54a514.